syntax = "proto3";

package hello.calibration;

// Represents pixel coordinates in an image
message ImagePoint {
  int32 x = 1;
  int32 y = 2;
  string tag = 3;
}

// Represents world coordinates in some sort of coordinate space. Typically the
// world is planar and z = 0.
message ObjectPoint {
  float x = 1;
  float y = 2;
  float z = 3;
  string tag = 4;
}

// Camera intrinsic matrix
message CameraMatrix {
  // Camera focal distance.
  // It is a product of the physical focal length of the lens and the size Sx
  // of the individual imager elements.
  // fx = F * Sx. Focal length in millimeters are not measurable
  float fx = 1;
  // The reason for fx is because the imagers are rectangular rather then square
  float fy = 2;
  // Center of projection on x to show possible displacement away from the
  // optical axis of the center of coordinate. Units are in pixel coordinates.
  float cx = 3;
  // Same as cx on a different axis.
  float cy = 4;
}

// Camera introduces distortion to images. During calibration the distortion
// parameters are identified. These parameters have two types:
// * Radial distortion causes lines to appear curved.
// * Tangential distortion occurs because lenses are not aligned
//   perfectly parallel to the imaging plane and the image appears nearer than
//   expected.
// See more at https://docs.opencv.org/4.x/dc/dbb/tutorial_py_calibration.html
message DistortionParameters {
  // There may be 4, 5 or 8 parameters which are 3 or more radial parameters
  // k1..k5 and 2 possible tangential parameters p1 and p2.
  // If length 4 then there should be k1, k2, p1 and p2.
  // If the length is 5 or 8, then the elements will be either (k1, k2, p1, p2
  // and k3) or (k1, k2, p1, p2, k3, k4, k5 and k6), respectively.
  // The five-element form is primarily for fisheye lenses.
  float k1 = 1;
  float k2 = 2;
  float k3 = 3;
  optional float k4 = 4;
  optional float k5 = 5;
  optional float p1 = 6;
  optional float p2 = 7;
}

// The result of homography which is a projective mapping from world plane to
// image plane.
// The matrix is 3x3 where two column vectors relate to rotational vectors
// and the last column is translational vector which is [r1, r2, t]
message HomographyMatrix {
  float r11 = 1;
  float r12 = 2;
  float t1 = 3;
  float r21 = 4;
  float r22 = 5;
  float t2 = 6;
  float r31 = 7;
  float r32 = 8;
  float t3 = 9;
}

// Rotation vector is a part of camera pose. It has radians in each components.
// For example [-0.7260306758687214, 1.965675002939307, -1.72538916801254]
// means [-42, 113, -99] => x * 180/3.14, etc.
message RotationVector {
  float x = 1;
  float y = 2;
  float z = 3;
}

// In some cases we need to serialize a number of object points into proto file.
message ObjectPoints {
  repeated ObjectPoint object_point = 1;
}

// This is handy to bulk import pixels rather than picking them in UI.
message ImagePoints {
  repeated ImagePoint image_point = 1;
}